### PR TITLE
Features/fixture decorator

### PR DESCRIPTION
**Pull request by sebastianmanger**
See https://github.com/behave/behave-django/issues/42

### CODE REVIEW

File: behave_django/decorators.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        for fixture in self.modules:
+            try:
+                import_string(fixture)(context)
+            except ImportError:
+                yield fixture
+
+
+def fixtures(*fixture_or_callable):
+    """
+    Provide fixtures for given step_imp. Fixtures will be loaded in
```
Comment (bittner) [inline behave_django/decorators.py]: Typo: "step_imp" is meant to read "step_impl".

**Comment by bittner** [comment]
@sebastianmanger Looks good for me.

@mixxorz Any comments?

**Comment by mixxorz** [comment]
This looks really good! Though, I have a couple of comments.

What is the purpose of the callables API? Is it merely for running arbitrary code before a step? If so, I feel like the fixtures decorator is not the right place for that. If this is the case, it's simpler to just call the function in the step body, or make the function into its own decorator.

Initially, I thought the callables was a way to dynamically return fixture strings, similar to how Django models accepts callables for [default](https://docs.djangoproject.com/en/1.11/ref/models/fields/#default). I feel like this use case makes more sense. What do you think?

**Comment by mixxorz** [comment]
In addition, I'd like to merge #44 first before this one because that PR fixes a bunch of issues with fixtures.

**Comment by sebastianmanger** [comment]
Thanks for the feedback!

* In reply to: "What is the purpose of the callables API"
I personally prefer 'dynamic' fixtures ([factory boy](http://factoryboy.readthedocs.io/en/latest/orms.html), [model mommy](http://model-mommy.readthedocs.io/en/latest/index.html) or such; sometimes also [VCR.py](https://vcrpy.readthedocs.io/en/latest/) for APIs) over 'static' json fixtures. While they are not django fixtures, they serve the same purpose, and I often need the same factory for multiple test cases. If you don't like the idea, I could provide two distinct decorators; one for fixtures, and one for callables. Or I could simply remove the option to provide callables.

* Right now, there could be duplicates in the context.fixtures list. I think I'll simply cast it to a set and then to a list again - commit comming soon.

**Comment by mixxorz** [comment]
I think I would be content if the return value of the callable is added to `context.fixtures` actually. You can still do whatever you want in it, but at least we can say it's purpose is to return a fixture name.

Though, I'm not so sure you can create database objects in there though. I am not positive that it's covered in the transaction.

Also please rebase to master.

File: docs/usage.rst [hunk 2]
```diff
@@ -147,6 +156,38 @@ mutate it however you want, it will only be processed upon leaving the
     meant to be used on the whole feature.  This is because Django's
     ``LiveServerTestCase`` resets the test database after each scenario.
 
+
+Fixtures using a decorator
+**************************
+
+You can define `django fixtures`_ or a path to a callable for any step, using
```
Comment (bittner) [inline docs/usage.rst]: I think, "Django" needs to be written with a capital initial letter. -- Unless you're talking about [django-fixtures](https://pypi.python.org/pypi/django-fixtures). :smiley:

**Comment by sebastianmanger** [comment]
I included callables in the decorator, as it was convenient for me - but thinking it through again, it seems it is not a clean approach. Coming from Django, it's not what you would expect from a 'fixture'. So if it's OK to you, I'll remove the option for callables (and do the other stuff as suggested).

File: behave_django/environment.py [hunk 3]
```diff
@@ -48,6 +49,17 @@ def load_fixtures(self, context):
         If fixtures are found in context, loads the fixtures using the loaddata
         management command.
         """
+        for step in context.scenario.all_steps:
+            match = step_registry.registry.find_match(step)
+            if match and hasattr(match.func, 'fixtures'):
+                if not hasattr(context, 'fixtures'):
+                    context.fixtures = []
+
+                # Load callables, and set fixture-strings to the context.
```
Comment (bittner) [inline behave_django/environment.py]: This comment should not be necessary. Obviously, the code below is somewhat too cryptic. Can we make it more self-explanatory? e.g.
```python
fixtures = match.func.fixtures.setup(context=context) or []
context.fixtures += fixtures
```
Though, I'm not really sure what the `fixtures` is that the `setup()` method returns. (What would be a self-explanatory name instead?)

Comment (sebastianmanger) [reply behave_django/environment.py]: Naming can and will be more concise once only fixtures (and not callables) are included in the decorator.

File: test_app/fixtures/callable.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+def my_callable(context):
+    context.test_me = 'hi!'
```
Comment (bittner) [inline test_app/fixtures/callable.py]: What does this callable do useful for the test?

Shouldn't it create an object (in the database)?

**Comment by bittner** [comment]
Let me add my two cents to the discussion.

1. I also prefer dynamic fixtures over static fixtures, as suggested by the [Django docs on testing](https://docs.djangoproject.com/en/stable/topics/testing/tools/#fixture-loading), mainly due to the high maintenance cost static fixtures incur.
1. I'm not sure how static fixtures can be generated from callables. Would be interesting to know.
1. Following PEP 20's idea of _"there should be one obvious way to do it"_, we should (at least) recommend one way over the other, similar to _"tests are more readable and it’s more maintainable to create objects using the ORM"_ in Django's testing docs. -- In the end, loading fixtures with annotations will result in the "creation" of the `fixtures` attribute on `context`. So, modifying `context.fixtures` would be "manipulating" the object part of the process directly, we could argue.

File: behave_django/environment.py (hunk 3 above)
Comment (bittner) [reply behave_django/environment.py]: That would make me sad, as I prefer callables over static fixtures (as you do).

File: behave_django/decorators.py [hunk 5]
```diff
@@ -1,43 +1,18 @@
...
+    environment.py#before_scenario.
 
-    @fixtures('mydata.json', 'path.to.my.callable')
+    @fixtures('data.json')
     @when('a user clicks the button')
     def step_impl(context):
         pass
 
-    :param fixture_or_callable: list of module-strings or fixture-files
+    :param fixture_files: list of fixture-files
```
Comment (bittner) [inline behave_django/decorators.py]: No dash between "fixture" and "files", please.

**Comment by sebastianmanger** [comment]
Current state: I hope I implemented all changes listed above.

'Callables' are currently not an option, as this seems to require some further discussion. I personally still prefer callables over static fixtures, but @mixxorz is right in saying that they can simply be called in the ``step_impl``. Also, "I'm not sure how static fixtures can be generated from callables. Would be interesting to know." from @bittner - I am not aware of any such tool/method. This leaves me with following options:

* no support for callables with the decorator, simply call them in the ``step_imp``
* a new decorator, specifically for callables
* (this is a new idea): fixtures as method arguments, as seen in [pytest fixtures](https://docs.pytest.org/en/latest/fixture.html)

Regarding maintenance, option number one seems preferable.

**Comment by mixxorz** [comment]
Nice work @sebastianmanger! This looks good to me.

About the callables, I think the biggest thing aside from API design is that I'm not sure if creating database objects in the same place where `context.fixtures` is mutated will actually work because I think that comes _before the test case is set up_. You might end up creating instances on your actual database.

Might not be fancy, but calling them from within the step function is better I think.

@bittner, does this look good to you?

**Comment by bittner** [comment]
Looks fine for me, too.

**TL;DR** -- Last question: Do we have to use `.json` at the end of the fixture, or does it also work without file extension, [as with Django](https://docs.djangoproject.com/en/stable/topics/testing/tools/#fixture-loading)?

What I care about is that we don't invent a "new thing". We should follow what Django does (and/or other popular tools do even better). In other words, the syntax should follow Django practices, it should be self-explanatory.

If that's clarified go ahead and merge, @mixxorz!

**Comment by bittner** [comment]
... and: @sebastianmanger get yourself a decent avatar! :smile: :boar:

**Comment by mixxorz** [comment]
The fixtures just append to the `TestCase.fixtures`, so it uses the same implementation as Django when loading them.

I'll go ahead and merge this!
